src/Workspaces/Core/Portable/Execution/Serialization/Asset.cs

```diff
@@ -0,0 +1,148 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Host;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Execution
+{
+    /// <summary>
+    /// Asset represents actual data. <see cref="HierarchicalChecksumObject"/> represents
+    /// collections of these data.
+    /// 
+    /// in hierarchical checksum tree, asset is leaf node, and hierarchicalChecksumObject
+    /// is node that has children
+    /// </summary>
+    internal abstract class Asset : ChecksumObject
+    {
+        public Asset(Checksum checksum, string kind) : base(checksum, kind)
+        {
+            // TODO: find out a way to reduce number of asset implementations.
+            //       tried once but couldn't figure out
+        }
+    }
+
+    /// <summary>
+    /// helper type for common assets
+    /// 
+    /// this asset will be used for data that require only 1 information
+    /// to serialize such as project reference, solution/project/document info and etc
+    /// </summary>
+    internal sealed class Asset<T> : Asset
+    {
+        private readonly T _value;
+        private readonly Action<T, ObjectWriter, CancellationToken> _writer;
+
+        public Asset(T value, string kind, Action<T, ObjectWriter, CancellationToken> writer) :
+            base(Checksum.Create(value, kind, writer), kind)
+        {
+            _value = value;
+            _writer = writer;
+        }
+
+        public override Task WriteToAsync(ObjectWriter writer, CancellationToken cancellationToken)
+        {
+            _writer(_value, writer, cancellationToken);
+            return SpecializedTasks.EmptyTask;
+        }
+    }
+
+    /// <summary>
+    /// helper type for language specific assets
+    /// 
+    /// this asset will be used for language specific data such as CompilationOption/ParseOption
+    /// </summary>
+    internal sealed class LanguageSpecificAsset<T> : Asset
```

Comment:
Basically, i'm trying to reduce complexity.The more types you have in these hierarchies, the less clear it is.  You already have a way to wrap an object in an asset and to provide a function to serialize it out.  So i don't know why this type is needed.   :)

Basically, i'd like us to ask ourselves with every part of this change "is this actually needed" and trim things that aren't.